### PR TITLE
remove additional rule that parameters must match for dispatch

### PR DESCRIPTION
**Pull request by vtjnash**
matching a method signature is a simple subtyping test,
with no additional rule to ensure that all parameters have a value

and adds a Test for possibly unbound parameters,
which allows detecting methods definitions that this change affects

### CODE REVIEW

File: base/dict.jl [hunk 1]
```diff
@@ -135,10 +135,10 @@ copy(d::Dict) = Dict(d)
 
 const AnyDict = Dict{Any,Any}
 
+Dict(ps::Pair...)                            = Dict{Any,Any}(ps) # define first, since method sort order is wrong for the following
```
Comment (kmsquire) [inline base/dict.jl]: Just curious, can you explain why the sort order is wrong when this is below?

Presumably, `DataStructures.jl` will need to make similar changes (e.g., at https://github.com/JuliaCollections/DataStructures.jl/blob/master/src/default_dict.jl#L88-L96)

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23117#issuecomment-320100959) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/cf0832654c58b168b40c8421a4dfa90479abe17a/af60994_vs_9b2a1cc/report.md). cc @ararslan

File: test/tuple.jl [hunk 2]
```diff
@@ -9,8 +9,59 @@ struct BitPerm_19352
...
 
+# ntuples
+nttest1(x::NTuple{n, Int}) where {n} = n
+@test nttest1(()) == 0
+@test nttest1((1, 2)) == 2
+@test NTuple <: Tuple
+@test (NTuple{T, Int32} where T) <: Tuple{Vararg{Int32}}
+@test !((NTuple{T, Int32} where T) <: Tuple{Int32, Vararg{Int32}})
+@test Tuple{Vararg{Int32}} <: (NTuple{T, Int32} where T)
+@test Tuple{Int32, Vararg{Int32}} <: (NTuple{T, Int32} where T)
```
Comment (vtjnash) [inline test/tuple.jl]: I'm not entirely sure why these tests weren't there. I assume there's already some in there? I just moved this block over from core.jl

File: base/essentials.jl [hunk 3]
```diff
@@ -159,11 +164,44 @@ function typename(a::Union)
...
-oftype(x,c) = convert(typeof(x),c)
+convert(::Type{T}, x::T) where {T<:Tuple{Any, Vararg{Any}}} = x
+convert(::Type{T}, x::Tuple{Any, Vararg{Any}}) where {T<:Tuple} =
+    (convert(tuple_type_head(T), x[1]), convert(tuple_type_tail(T), tail(x))...)
+
+# TODO: the following definitions are equivalent (behaviorally) to the above method
+# I think they may be faster / more efficient for inference,
+# if we could enable them, but are they?
+# TODO: These currently can't be used (#21026) since with
+#     z(::Type{Tuple{Val{T}} where T}) = T
```
Comment (vtjnash) [inline base/essentials.jl]: Indeed, I seem to have been tired when I tried to write the MRE. ref https://github.com/JuliaLang/julia/issues/23017

On master, this bug in the subtyping environment exhibits itself as a MethodError:
```julia
julia> z2(::Type{<:Tuple{Vararg{T}}}) where {T} = T

julia> z2(Tuple{Val{T}} where T)
Val{T}

julia> z2(Tuple{Val})
ERROR: MethodError: no method matching z2(::Type{Tuple{Val}})
```

File: src/subtype.c [hunk 4]
```diff
@@ -2583,6 +2585,46 @@ JL_DLLEXPORT int jl_type_morespecific_no_subtype(jl_value_t *a, jl_value_t *b)
...
+        jl_svec_t *e = jl_alloc_svec(szb);
+        *penv = e;
+        for (i = 0; i < szb; i++)
+            jl_svecset(e, i, env[i]);
+    }
+    JL_GC_POP();
+    return sub;
+}
+
+JL_DLLEXPORT jl_value_t *jl_match_subtype(jl_value_t *type, jl_value_t *sig)
```
Comment (tkelman) [inline src/subtype.c]: is this called anywhere, does it need to be exported?

File: src/typemap.c [hunk 5]
```diff
@@ -683,9 +649,9 @@ static jl_typemap_entry_t *jl_typemap_lookup_by_type_(jl_typemap_entry_t *ml, jl
 
 
 // this is the general entry point for looking up a type in the cache
-// (as a subtype, or with typeseq)
+// as a subtype, or with type_equal
```
Comment (tkelman) [inline src/typemap.c]: is this now referring to `jl_types_equal` instead of the Julia `Base.typeseq` ?

File: test/tuple.jl [hunk 6]
```diff
@@ -9,8 +9,59 @@ struct BitPerm_19352
...
+    @test convert(Tuple{Vararg{Int}}, ()) == ()
+    @test convert(Tuple{Vararg{T}} where T<:Integer, ()) == ()
+
+    @test convert(NTuple{3, Int}, (1, 2, 3)) == (1, 2, 3)
+    @test convert(NTuple, (1, 2, 3)) == (1, 2, 3)
+    @test convert(Tuple{Vararg{Int}}, (1, 2, 3)) == (1, 2, 3)
+    @test convert(Tuple{Int, Vararg{Int}}, (1, 2, 3)) == (1, 2, 3)
+    @test convert(Tuple{Vararg{T}} where T<:Integer, (1, 2, 3)) == (1, 2, 3)
+    @test convert(Tuple{T, Vararg{T}} where T<:Integer, (1, 2, 3)) == (1, 2, 3)
+    @test convert(Tuple{Int, Int, Float64}, (1, 2, 3)) == (1, 2, 3.0)
```
Comment (tkelman) [inline test/tuple.jl]: this (and the rest here too?) should probably be using `===`

**Review by tkelman — COMMENTED** [review]
few minor things, can be addressed later if needed

**Comment by JeffBezanson** [comment]
Benchmarks look generally good. Why the huge speedup in "splatting"?

**Comment by KristofferC** [comment]
> Benchmarks look generally good

Hmm, 6 of the `problem` benchmarks which are quite large show regressions.

File: src/subtype.c [hunk 7]
```diff
@@ -2583,6 +2585,46 @@ JL_DLLEXPORT int jl_type_morespecific_no_subtype(jl_value_t *a, jl_value_t *b)
     return type_morespecific_(a, b, 0, NULL);
 }
 
+JL_DLLEXPORT jl_svec_t *jl_match_method(jl_value_t *a, jl_value_t *b)
```
Comment (JeffBezanson) [inline src/subtype.c]: This function is identical to `jl_env_from_type_intersection`; let's remove it. (`jl_env_from_type_intersection` looks like it's missing a GC root though)

File: base/tuple.jl [hunk 8]
```diff
@@ -63,6 +63,7 @@ first(t::Tuple) = t[1]
 # eltype
 
 eltype(::Type{Tuple{}}) = Bottom
+#eltype(::Type{Tuple{Vararg{E}}}) where {E} = E
```
Comment (JeffBezanson) [inline base/tuple.jl]: This could use a comment or `TODO` note, or delete it.

File: src/codegen.cpp [hunk 9]
```diff
@@ -3064,7 +3064,12 @@ static jl_cgval_t emit_sparam(jl_codectx_t &ctx, size_t i)
             T_prjlvalue,
             ctx.spvals_ptr,
             i + sizeof(jl_svec_t) / sizeof(jl_value_t*));
-    return mark_julia_type(ctx, tbaa_decorate(tbaa_const, ctx.builder.CreateLoad(bp)), true, jl_any_type, false);
+    Value *sp = tbaa_decorate(tbaa_const, ctx.builder.CreateLoad(bp));
+    Value *isnull = ctx.builder.CreateICmpNE(emit_typeof(ctx, sp),
```
Comment (JeffBezanson) [inline src/codegen.cpp]: Looks like this variable should be called `isdef` instead?

File: src/subtype.c (hunk 7 above)
Comment (JeffBezanson) [inline src/subtype.c]: And since it returns a pair it should be renamed e.g. `jl_type_intersection_with_env`.

**Review by JeffBezanson — APPROVED** [review]

**Comment by KristofferC** [comment]
I really thing the `problem` benchmarks should be discussed more.

**Comment by ararslan** [comment]
We can check if they persist in another Nanosoldier run: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by JeffBezanson** [comment]
Here's what I see in the graphs:

ziggurat: Gradually got worse since Jan 1.
raytrace: big spike early July
stockcorr: Got worse starting around Jan 1, got better in May.
imdb: Has been improving
grigoriadis: Very similar curve to ziggurat

It definitely seems worth investigating what happened starting in the Jan-Mar timeframe, but it's not clear to me if any of this relates to this PR.

**Comment by KristofferC** [comment]
It relates to the PR in that a 20% bump for some of these problems would make them right now the slowest they have ever been since we started recording it.

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23117#issuecomment-320338860) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8a662312a24d2dab0762d9f7648b5e86cdf37007/b0fdea7_vs_45144d8/report.md). cc @ararslan

**Comment by JeffBezanson** [comment]
Looks like raytracing still shows a slowdown but the other slowdowns in the "problem" benchmarks went away.

**Comment by tkelman** [comment]
nullablearray and one of the "random" benchmarks also show a consistent more-than-2x slowdown

File: src/typemap.c (hunk 5 above)
Comment (vtjnash) [inline src/typemap.c]: same thing

File: src/codegen.cpp (hunk 9 above)
Comment (vtjnash) [inline src/codegen.cpp]: That would probably have been a more logical choice. Currently, I'm just maintaining compatibility with the handful of other places we got this backwards, and leaving fixing the names to a future PR author.

File: src/typemap.c (hunk 5 above)
Comment (tkelman) [inline src/typemap.c]: sure but now the comment isn't saying either of those

Comment (vtjnash) [inline src/typemap.c]: I meant that all three are the same statement

Comment (tkelman) [inline src/typemap.c]: this is the only place in the entire repo that "type_equal" appears

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23117#issuecomment-320810757) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/dd41e668b8ef9e56361d9bb88cd1fff4f044c0a6/79619de_vs_4fea203/report.md). cc @ararslan

**Comment by vtjnash** [comment]
Looks like everything is good here now. OK to merge?

**Comment by JeffBezanson** [comment]
Yep, ok by me.

**Comment by StefanKarpinski** [comment]
I'm more excited about this change than I have any right to be.

File: NEWS.md [hunk 10]
```diff
@@ -54,6 +54,15 @@ Language changes
...
     contains `sin` eagerly, rather than delaying that decision until `f` is run. ([#22984]).
 
+  * Dispatch rules have been simplified:
+    matching methods is now determined exclusively by subtyping;
+    the rule that method type parameters must be also be captured has been removed.
+    Instead, attempting to access the uncontrained parameters will throw an `UndefVarError`.
+    Linting in package tests is recommended to confirm that the set of methods
+    which might throw `UndefVarError` when accessing the static parameters
+    (`need_to_handle_undef_sparam = Set{Any}(m.sig for m in Test.detect_unbound_args(Base, recursive=true))`)
+    is equal (`==`) to some known set (`expected = Set()`). ([#TBD])
```
Comment (ararslan) [inline NEWS.md]: Should update the TBD here

File: base/dict.jl (hunk 1 above)
Comment (kmsquire) [reply base/dict.jl]: Bump, @vtjnash.

Comment (vtjnash) [reply base/dict.jl]: Probably fixed now—Jeff has made a lot of changes to method sort order since then

**Pull request merged**